src/Http/ServerRequest.php

```diff
@@ -577,16 +584,54 @@ public function session(Session $session = null)
      */
     public function clientIp()
     {
-        if ($this->trustProxy && $this->getEnv('HTTP_X_FORWARDED_FOR')) {
-            $addresses = explode(',', $this->getEnv('HTTP_X_FORWARDED_FOR'));
-            $ipaddr = end($addresses);
-        } elseif ($this->trustProxy && $this->getEnv('HTTP_CLIENT_IP')) {
-            $ipaddr = $this->getEnv('HTTP_CLIENT_IP');
-        } else {
-            $ipaddr = $this->getEnv('REMOTE_ADDR');
+        if ($this->trustProxy) {
+            if ($forwarded = $this->getEnv('HTTP_X_FORWARDED_FOR')) {
+                $addresses = array_map('trim', explode(',', $forwarded));
+                $trusted = (count($this->trustedProxies) > 0);
+                $n = count($addresses);
+
+                if ($trusted) {
+                    for ($i = 1; $i < $n; $i++) {
+                        if (in_array($addresses[$i], $this->trustedProxies) === false) {
+                            $trusted = false;
+                            break;
+                        }
+                    }
+                }
```

Comment:
This looks like it could be an `array_diff`?

Reply:
no there maybe more trusted proxies than X FORWARDED_FOR

Reply:
That should be ok as `array_diff($addresses, $this->trustedProxies)` would give you the addresses that are not trusted. If that list isn't empty `$trusted` becomes false.

Reply:
yes now array_diff added :smile: